contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1405 @@
+            uint256 poolBalance = IERC20(_listTCO2[i]).balanceOf(_toucanPool);
+
+            // Error check for possible 0 balance / stale lists
+            if (poolBalance == 0) {} else {
+                address[] memory redeemERC20 = new address[](1);
+                redeemERC20[0] = _listTCO2[i];
+
+                uint256[] memory redeemAmount = new uint256[](1);
+
+                // Burn only pool balance if there are more pool tokens than available
+                if (leftToBurn > poolBalance) {
+                    // Redeem from pool
+                    redeemAmount[0] = poolBalance;
+                    IBaseCarbonTonne(_toucanPool).redeemMany(
+                        redeemERC20,
+                        redeemAmount
+                    );
+
+                    // Retire TCO2
+                    IToucanCarbonOffsets(_listTCO2[i]).retire(poolBalance);
+                    emit ToucanRetired(
+                        msg.sender,
+                        _beneficiary,
+                        _toucanPool,
+                        _listTCO2[i],
+                        poolBalance
+                    );
+
+                    leftToBurn -= poolBalance;
+                } else {
+                    // Redeem from pool
+                    redeemAmount[0] = leftToBurn;
+                    IBaseCarbonTonne(_toucanPool).redeemMany(
+                        redeemERC20,
+                        redeemAmount
+                    );
+
+                    // Retire TCO2
+                    IToucanCarbonOffsets(_listTCO2[i]).retire(leftToBurn);
+                    emit ToucanRetired(
+                        msg.sender,
+                        _beneficiary,
+                        _toucanPool,
+                        _listTCO2[i],
+                        leftToBurn
+                    );
+
+                    leftToBurn = 0;
+                }
+            }
+        }
+
+        require(
+            leftToBurn == 0,
+            "Not all pool tokens were burned. Please submit a larger TCO2 token list."
+        );
+    }
+
+    function _bondFees(address _toucanPool) internal returns (bool) {
+        // Bond cumulative fees to the DAO if bond is large enough
+        uint256 poolBalance = IERC20(_toucanPool).balanceOf(address(this));
+
+        uint256 bondPrice = IKlimaBondDepository(bondDepository[_toucanPool])
+            .bondPriceInUSD();
+
+        if ((poolBalance * (10**9)) / bondPrice > 10000000) {
+            IERC20(_toucanPool).safeIncreaseAllowance(
+                bondDepository[_toucanPool],
+                poolBalance
+            );
+
+            IKlimaBondDepository(bondDepository[_toucanPool]).deposit(
+                poolBalance,
+                bondPrice,
+                DAO
+            );
+        }
+        return true;
+    }
+
+    function setFeeAmount(uint256 _amount) external onlyOwner returns (bool) {
+        emit FeeUpdated(feeAmount, _amount);
```

Comment:
Nitpicking but I would recommend to establish a pattern where events are emitted after state changes. It doesn't necessarily apply here but as a general rule: https://twitter.com/transmissions11/status/1473205292691296256

Reply:
I think I got everything set up with some additional events to emit them after any state changes in the contract showing previous and new values.